### ISSUE TITLE
IllegalStateException in WebSocket ConnectionState

### ISSUE THREAD

**Issue by nxmbriggs404** [eclipse/jetty.project#3476]
We have a jetty based project that we've recently attempted to upgrade to jetty 9.4.15.v20190215 from 9.2.25.v20180606.  When attempting to connect a WebSocket via WebSocketClient.connect, we're seeing occasional errors in testing.  Unfortunately I don't know the jetty code base very well, but on the surface it looks to me as if it an internal invariant is being violated (i.e. it's not a normal sort of I/O error).  I wonder if it's a race condition and an connection/session/endpoint/etc... might be getting reused inadvertently?  Complete speculation on my part.

Please find the stack trace that we're capturing during testing below.

I'm in a good position to help dig deeper if there are any hints as to how to proceed.

Thanks!

```java
  2019-03-18 16:46:43,904	76348   	ERROR 	senger-Sessions	Error connecting to messenger; messengerUri=ws://127.0.0.1:25142/messenger	[]	c.n.j.m.w.ConnectFailureLogger
  java.lang.IllegalStateException: ConnectionState@24d90109[OPENING]
	at org.eclipse.jetty.websocket.common.io.ConnectionState.opening(ConnectionState.java:76) ~[websocket-common-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.opening(AbstractWebSocketConnection.java:253) ~[websocket-common-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.websocket.common.WebSocketSession.open(WebSocketSession.java:441) ~[websocket-common-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.websocket.common.WebSocketSession.onOpened(WebSocketSession.java:415) ~[websocket-common-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.io.AbstractConnection.onOpened(AbstractConnection.java:209) ~[jetty-io-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.io.AbstractConnection.onOpen(AbstractConnection.java:202) ~[jetty-io-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.io.AbstractEndPoint.upgrade(AbstractEndPoint.java:441) ~[jetty-io-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.websocket.client.WebSocketUpgradeRequest.upgrade(WebSocketUpgradeRequest.java:636) ~[websocket-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.exchangeTerminating(HttpChannelOverHTTP.java:118) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:462) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.HttpReceiver.responseSuccess(HttpReceiver.java:416) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.messageComplete(HttpReceiverOverHTTP.java:316) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.http.HttpParser.handleHeaderContentMessage(HttpParser.java:591) ~[jetty-http-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.http.HttpParser.parseFields(HttpParser.java:1213) ~[jetty-http-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1502) ~[jetty-http-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:172) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:135) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155) ~[jetty-client-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305) ~[jetty-io-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) ~[jetty-io-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) ~[jetty-io-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) ~[jetty-util-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) ~[jetty-util-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) ~[jetty-util-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:132) ~[jetty-util-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765) ~[jetty-util-9.4.15.v20190215.jar:9.4.15.v20190215]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683) ~[jetty-util-9.4.15.v20190215.jar:9.4.15.v20190215]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_192]
```

**Comment by sbordet** [comment]
Seems indeed a race of some kind.

Can you reproduce with DEBUG logs enabled and attach the logs to this issue?

**Comment by nxmbriggs404** [comment]
Hi Simone, thank you for the quick reply!

I haven't yet attempted to ratchet up the jetty logs to DEBUG, but I was considering it.  I'll go ahead and do that and hopefully we'll still get a repro.

**Comment by joakime** [comment]
It's a strange error/issue.
That stacktrace means that the WebSocket was opened twice??
Running in debug when this happens is going to be the best way to understand what happened. (or a reproducer use case / test case)
If you cannot run with full debugging, then please limit debug to the named logger `org.eclipse.jetty.websocket`, and hopefully there will be enough to work with.

**Comment by nxmbriggs404** [comment]
I was able to run with full jetty DEBUG logging enabled, but unfortunately I didn't get a repro this time.  I suspect the log volume tweaked the timing enough to avoid the condition.

I will give it another go with the targeted org.eclipse.jetty.websocket DEBUG logging instead!

**Comment by nxmbriggs404** [comment]
I ran our test suite four times with org.eclipse.jetty.websocket logging set to DEBUG, and unfortunately the issue didn't repro during any of those runs.

I reran the suite with all jetty logging set back to default levels, and the issue reproduced.

**Comment by joakime** [comment]
@nxmbriggs404 do you have a project or small use case that can replicate this?
Perhaps you can limit the debug logging even further.  set only `org.eclipse.jetty.websocekt.common.WebSocketSession` to DEBUG?

**Comment by joakime** [comment]
Just a thought, since this appears to be happening in `WebSocketSession.open()` you can attach a `org.eclipse.jetty.websocket.common.WebSocketSessionListener` to the `WebSocketClient.addSessionListener()` to listen for all open/closed Sessions.
If you can setup some code to track this, you might be able to tell what is happening.
You'll see open/close for each WebSocketSession, and once the fatal `java.lang.IllegalStateException: ConnectionState@24d90109[OPENING]` you should be be able to track down what connection/request caused this issue.
Alternatively, I could create an instrumented version of the websocket-common.jar to add more tracking of the states in the hopes of tracking down how this is happening?

**Comment by nxmbriggs404** [comment]
Thank you for the follow up and suggestions.

I've been experimenting with jetty 9.3.25.v20180904 as an alternative for us to 9.4, and it's actually been consistently passing all of our tests.  It also looks like this 9.3 version will satisfy our immediate upgrade goals, so the pressure for us to migrate to 9.4 is less.  I would still like to help get to the bottom of our test failures with 9.4, though.  My hope is to come back to this in the next couple of weeks to work through the suggestions.

Thanks again!

**Comment by joakime** [comment]
jetty-9.3.x is in maintenance mode at the moment.
jetty-9.4.x is current stable branch (on Java 8, and likely be supported for years.  depends on community interest in Java 8 ongoing)
jetty-10.0.x is the upcoming release branch. (on Java 11, is our future stable branch, this one has a large refactor of the core/internal websocket implementation btw)

**Comment by kan-izh** [comment]
I think I have similar problem. My suspicion it happens when I do session.getRemote().sendString right from onWebSocketConnect callback. I think some `volatile` are missing somewhere inside Jetty.

**Comment by joakime** [comment]
@kan-izh the move to state `opening` happens before the application's `onWebSocketConnect` is even called.
But I'll happily test out this theory.

**Comment by kan-izh** [comment]
Yeah... Logically seems you are right But it "happens before" by the code, and it does not necessary means that it "happens before" in the sense of Java Memory Model.

**Comment by joakime** [comment]
@kan-izh are you using the websocket layer in jetty normally? or via spring?

**Comment by kan-izh** [comment]
Normal `new WebSocketClient()` with WebSocketListener. Just hammering it with many-many connect/requests/responses/close
BTW, also I occasionally see the following exceptions, but they are not fatal, because happening at the very end:
```
java.nio.channels.ClosedChannelException: null
	at org.eclipse.jetty.io.WriteFlusher.onClose(WriteFlusher.java:502)
	at org.eclipse.jetty.io.AbstractEndPoint.onClose(AbstractEndPoint.java:353)
	at org.eclipse.jetty.io.ChannelEndPoint.onClose(ChannelEndPoint.java:215)
	at org.eclipse.jetty.io.AbstractEndPoint.doOnClose(AbstractEndPoint.java:225)
	at org.eclipse.jetty.io.AbstractEndPoint.close(AbstractEndPoint.java:192)
	at org.eclipse.jetty.io.AbstractEndPoint.close(AbstractEndPoint.java:175)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.disconnect(AbstractWebSocketConnection.java:328)
	at org.eclipse.jetty.websocket.common.io.DisconnectCallback.failed(DisconnectCallback.java:36)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.close(AbstractWebSocketConnection.java:200)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.remoteClose(AbstractWebSocketConnection.java:279)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:109)
	at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:319)
	at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:201)
	at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:221)
	at org.eclipse.jetty.websocket.common.Parser.parse(Parser.java:246)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.readParse(AbstractWebSocketConnection.java:577)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:434)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by kan-izh** [comment]
I am looking at the code and have some doubts:
```java
    public void open()
    {
        if(LOG.isDebugEnabled())
            LOG.debug("[{}] {}.open()",policy.getBehavior(),this.getClass().getSimpleName());
// !!! Another thread is reading this and sees "null" even if it was already assigned by other thread
        if (remote != null)
        {
            // already opened
            return;
        }

        try(ThreadClassLoaderScope scope = new ThreadClassLoaderScope(classLoader))
        {
            // Upgrade success
 // !!! It is threadsafe, ok.
            if(connection.opening())
            {
                // Connect remote
 // !!! This is non-volatile variable, so it may be not membared to another thread
                remote = remoteEndpointFactory.newRemoteEndpoint(connection, outgoingHandler, getBatchMode());
```

**Comment by joakime** [comment]
@kan-izh good analysis, but i'd be more concerned with the fact that you think there would be more then 1 thread accessing the same WebSocketSession at this point in time.

At the point of `WebSocketSession.open()` the connection is single threaded, working its way through the upgrade process, which called the internal jetty `Connection.onOpen()`, that triggered the various `Connection.Listener`s of which `WebSocketSession` is a part of.   

On a server side WebSocket, this is pretty much guaranteed to be single threaded.  On a client side, this is ... uhm ... @sbordet what's the possibility here of multiple threads?

**Comment by kan-izh** [comment]
My understanding is.... when the connection thread working through the upgrade process, it invokes `onWebSocketConnect` along the way, but not at the very end. And I have `sendString()` which works asynchronously, i.e. goes into another thread to send a frame - and the race begins! Plausible?

**Comment by kan-izh** [comment]
I've set a small project for a ping-pong ws client-server. Every run I see multiple exceptions including the `IllegalStateException`.
Please review the code. Maybe I use Jetty api not how it supposed to be. Or use it as a way to reproduce the issue.
https://github.com/kan-izh/jetty3476

**Comment by lachlan-roberts** [comment]
@joakime the `IllegalStateException` looks like it might be a problem with the `Connection.Listener` mechanism

in `WebSocketUpgradeRequest.upgrade()`
any `Connection.Listener` beans on the `WebSocketClient` will be added as listeners on the `WebSocketClientConnection`, then later in this method the `WebSocketSession` is added as a bean onto the the client. So when the next connection is to be opened, the session from the previous connection will also be added as a listener on the new connection as `WebSocketSession` implements `Connection.Listener`.

**Comment by joakime** [comment]
PR #3557 is merged. (and available starting in `jetty-9.4.18.v20190429`)
